lib/clib/Corres_UL_C.thy

```diff
@@ -1719,4 +1719,113 @@ lemma ccorres_grab_asm:
    ccorres_underlying sr G rr xf ar ax (P and K Q) P' hs f g"
   by (fastforce simp: ccorres_underlying_def)
 
+
+\<comment> \<open> An experimental cong rule for rewriting everywhere reasonable, with full context.
+    Can cause problems when there are schematic variables or when one of the return relations
+    takes a pair as a parameter. \<close>
+lemma ccorres_context_cong_helper':
+  assumes c: "ccorres_underlying sr \<Gamma> r xf ar axf P Q hs a c"
+  assumes "\<And>s. P s = P' s"
+  \<comment> \<open>Don't add context for rewriting Q, as the LHS can be simplified into something that is
+     unable to unify with the RHS.
+  assumes "\<And>s s'. \<lbrakk> (s,s') \<in> sr; P' s \<rbrakk> \<Longrightarrow> s' \<in> Q = (s' \<in> Q')"\<close>
+  assumes "Q = Q'"
```

Comment:
If we wanted to, we could actually still add a little bit of context:
```
assumes "\<And>s s'. \<lbrakk> (s,s') \<in> sr; P' s \<rbrakk> \<Longrightarrow> Q = Q'"
```
This would allow it to use predicates or parts of the state relation that don't depend on the state in the simplification of `Q`.

Reply:
Good idea, I hadn't thought of parts that didn't depend on the state.
I've done this now and it at least didn't break the one or two places these cong rules are currently being used.